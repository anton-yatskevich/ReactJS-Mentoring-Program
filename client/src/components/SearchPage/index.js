import React from 'react';
import PropTypes from 'prop-types';
import ResultsList from '../../containers/ResultsListContainer';
import SearchPanel from '../../containers/SearchPanel';
import SearchDescription from '../SearchDescription';
import './styles.scss';

const SearchPage = ({ numberOfResults }) => (
    <>
        <SearchPanel />
        <SearchDescription numberOfResults={numberOfResults} />
        <ResultsList />
    </>
);

SearchPage.propTypes = {
    numberOfResults: PropTypes.number
};

SearchPage.defaultProps = {
    numberOfResults: 0
};

export default SearchPage;
