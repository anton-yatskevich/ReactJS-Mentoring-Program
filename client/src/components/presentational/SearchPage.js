import React, { Component } from 'react';
import Foooter from './Footer';
import Header from './Header';
import ResultsList from './Results';

class SearchPage extends Component {
	constructor() {
		super();
		this.state = {
			results: [
				{key: 1, text: 'First item'},
				{key: 2, text: 'Second item'},
				{key: 3, text: 'Third item'},
				{key: 4, text: 'Fourth item'}
			]
		};
	}

	render() {
		return (
			<>
				<Header title='netflixroulette' />
				<ResultsList {...this.state} />
				<Foooter />
			</>
		)
	}
}

export default SearchPage;
