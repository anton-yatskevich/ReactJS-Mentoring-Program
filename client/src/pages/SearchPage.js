import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import noop from 'lodash.noop';
import ReactRouterPropTypes from 'react-router-prop-types';
import Header from '../components/Header';
import Footer from '../components/Footer';
import ResultsList from '../containers/ResultsListContainer';
import SearchPanel from '../containers/SearchPanel';
import SearchDescription from '../components/SearchDescription';
import setSearchQuery from '../actions/setSearchQuery';
import fetchMovies from '../actions/fetchMovies';

class SearchPage extends Component {
    componentDidMount() {
        const { match: { params }, setQuery, getMovies } = this.props;
        const query = params.query || '';
        setQuery(query);
        getMovies();
    }

    render() {
        const { numberOfResults } = this.props;
        return (
            <>
                <Header />
                <main>
                    <SearchPanel />
                    <SearchDescription numberOfResults={numberOfResults} />
                    <ResultsList />
                </main>
                <Footer />
            </>
        );
    }
}

SearchPage.propTypes = {
    match: ReactRouterPropTypes.match.isRequired,
    setQuery: PropTypes.func,
    getMovies: PropTypes.func,
    numberOfResults: PropTypes.number
};

SearchPage.defaultProps = {
    setQuery: noop,
    getMovies: noop,
    numberOfResults: 0
};

function mapStateToProps({ movies, searchParams }) {
    return {
        numberOfResults: movies.moviesList.length,
        searchQuery: searchParams.searchQuery
    };
}

export default withRouter(
    connect(mapStateToProps, { setQuery: setSearchQuery, getMovies: fetchMovies })(SearchPage)
);
