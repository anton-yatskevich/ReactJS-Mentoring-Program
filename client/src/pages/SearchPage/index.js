import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import noop from 'lodash.noop';
import ReactRouterPropTypes from 'react-router-prop-types';
import ResultsList from '../../containers/ResultsListContainer';
import SearchPanel from '../../containers/SearchPanel';
import SearchDescription from '../../components/SearchDescription';
import setSearchQuery from '../../actions/setSearchQuery';
import fetchMovies from '../../actions/fetchMovies';

class SearchPage extends Component {
    static fetchData({ dispatch }) {
        return dispatch(fetchMovies());
    }

    componentDidMount() {
        // const { match: { params }, setQuery, getMovies } = this.props;
        // const query = params.query || '';
        // setQuery(query);
        // getMovies();
    }

    render() {
        const { numberOfResults } = this.props;
        return (
            <>
                <SearchPanel />
                <SearchDescription numberOfResults={numberOfResults} />
                <ResultsList />
            </>
        );
    }
}

SearchPage.propTypes = {
    setQuery: PropTypes.func,
    getMovies: PropTypes.func,
    numberOfResults: PropTypes.number
};

SearchPage.defaultProps = {
    setQuery: noop,
    getMovies: noop,
    numberOfResults: 0
};

function mapStateToProps({ movies, searchParams }) {
    return {
        numberOfResults: movies.moviesList.length,
        searchQuery: searchParams.searchQuery
    };
}
export default connect(
    mapStateToProps,
    { setQuery: setSearchQuery, getMovies: fetchMovies }
)(SearchPage);
