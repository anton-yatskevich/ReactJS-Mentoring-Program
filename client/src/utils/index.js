import moment from 'moment';

function filterDatesHelper(a, b) {
    const momentA = moment(a.release_date, 'YYYY-MM-DD');
    const momentB = moment(b.release_date, 'YYYY-MM-DD');

    return momentB - momentA;
}

export const sortMoviesComporator = (a, b, sortField) => {
    switch (sortField) {
    case 'rating':
        return b.vote_average - a.vote_average;
    case 'release date':
        return filterDatesHelper(a, b);
    default:
        return 0;
    }
};

export const filterMoviesComporator = (movie, searchValue, searchField) => {
    const field = movie[searchField];

    if (Array.isArray(field)) {
        return field.some(item => item.toLowerCase().indexOf(searchValue) !== -1);
    }
    return field.toLowerCase().indexOf(searchValue) !== -1;
};
