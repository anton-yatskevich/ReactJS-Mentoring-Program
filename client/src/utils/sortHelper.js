import moment from 'moment';

function filterDatesHelper(a, b) {
    const momentA = moment(a.release_date, 'YYYY-MM-DD');
    const momentB = moment(b.release_date, 'YYYY-MM-DD');

    return momentB - momentA;
}

export default (movies, sortField) => (
    [...movies].sort((a, b) => {
        switch (sortField) {
        case 'rating':
            return b.vote_average - a.vote_average;
        case 'release_date':
            return filterDatesHelper(a, b);
        default:
            return 0;
        }
    })
);
