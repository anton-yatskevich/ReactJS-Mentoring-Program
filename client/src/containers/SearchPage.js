import React, { Component } from 'react';
import PropTypes from 'prop-types';

import ResultsList from '../components/ResultsList';
import SearchPanel from './SearchPanel';
import SearchDescription from '../components/SearchDescription';

import { sortMoviesComporator, filterMoviesComporator } from '../utils';

class SearchPage extends Component {
    constructor(props) {
        super(props);
        this.movies = props.movies;
        this.onSearchFormSubmit = this.onSearchFormSubmit.bind(this);
        this.onChangeSearchField = this.onChangeSearchField.bind(this);
        this.onChangeSort = this.onChangeSort.bind(this);
        this.state = {
            results: [],
            searchField: 'title',
            sortField: 'rating'
        };
    }

    onSearchFormSubmit(value) {
        const { searchField, sortField } = this.state;

        this.setState({
            results: this.movies
                .filter(movie => filterMoviesComporator(movie, value, searchField))
                .sort((a, b) => sortMoviesComporator(a, b, sortField))
        });
    }

    onChangeSearchField(value) {
        this.setState({
            searchField: value
        });
    }

    onChangeSort(sortField) {
        const { results } = this.state;

        this.setState({
            sortField,
            results: results
                .sort((a, b) => sortMoviesComporator(a, b, sortField))
        });
    }

    render() {
        const { searchField, sortField, results } = this.state;
        const { onSelectMovie } = this.props;

        return (
            <>
                <SearchPanel
                    searchField={searchField}
                    onSubmit={this.onSearchFormSubmit}
                    onChangeSearchField={this.onChangeSearchField}
                />
                <SearchDescription
                    numberOfResults={results.length}
                    sortField={sortField}
                    onChangeSort={this.onChangeSort}
                />
                <ResultsList movies={results} onSelectMovieHandler={onSelectMovie} />
            </>
        );
    }
}

SearchPage.propTypes = {
    movies: PropTypes.arrayOf(PropTypes.objectOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
            PropTypes.arrayOf(PropTypes.string)
        ])
    )),
    onSelectMovie: PropTypes.func
};

SearchPage.defaultProps = {
    movies: [],
    onSelectMovie: null
};

export default SearchPage;
