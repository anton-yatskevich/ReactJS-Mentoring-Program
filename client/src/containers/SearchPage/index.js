import React, { Component } from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash.noop';

import ResultsList from '../../components/ResultsList';
import SearchPanel from '../SearchPanel';
import SearchDescription from '../../components/SearchDescription';
import { sortMoviesComporator, filterMoviesComporator } from '../../utils';
import './styles.scss';

class SearchPage extends Component {
    state = {
        results: [],
        searchField: 'title',
        sortField: 'rating'
    };

    onSearchFormSubmit = (value) => {
        const { searchField, sortField } = this.state;
        const { movies } = this.props;

        this.setState({
            results: movies
                .filter(movie => filterMoviesComporator(movie, value, searchField))
                .sort((a, b) => sortMoviesComporator(a, b, sortField))
        });
    }

    onChangeSearchField = (value) => {
        this.setState({
            searchField: value
        });
    }

    onChangeSort = (sortField) => {
        const { results } = this.state;

        this.setState({
            sortField,
            results: results
                .sort((a, b) => sortMoviesComporator(a, b, sortField))
        });
    }

    render() {
        const { searchField, sortField, results } = this.state;
        const { onSelectMovie } = this.props;

        return (
            <>
                <SearchPanel
                    searchField={searchField}
                    onSubmit={this.onSearchFormSubmit}
                    onChangeSearchField={this.onChangeSearchField}
                />
                <SearchDescription
                    numberOfResults={results.length}
                    sortField={sortField}
                    onChangeSort={this.onChangeSort}
                />
                <ResultsList movies={results} onSelectMovieHandler={onSelectMovie} />
            </>
        );
    }
}

SearchPage.propTypes = {
    movies: PropTypes.arrayOf(PropTypes.objectOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
            PropTypes.arrayOf(PropTypes.string)
        ])
    )),
    onSelectMovie: PropTypes.func
};

SearchPage.defaultProps = {
    movies: [],
    onSelectMovie: noop
};

export default SearchPage;
