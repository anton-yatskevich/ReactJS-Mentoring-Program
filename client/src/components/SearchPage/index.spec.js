import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import SearchPage from './index';
import * as utils from '../../utils/sortHelper';

jest.mock('../../utils', () => ({
    sortMoviesComporator: jest.fn(() => []),
    filterMoviesComporator: jest.fn(() => [])
}));


describe('SearchPage component', () => {
    let SearchPageComponent;

    beforeAll(() => {
        SearchPageComponent = shallow(<SearchPage />);
    });

    it('should be rendered with default props', () => {
        expect(toJson(SearchPageComponent)).toMatchSnapshot();
    });

    it('should set searchField value if onChangeSearchField is called', () => {
        SearchPageComponent.instance().onChangeSearchField('testValue');
        expect(SearchPageComponent.state().searchField).toBe('testValue');
    });

    it('should set searchField value and sort results if onChangeSort method is called', () => {
        SearchPageComponent.state().results = [{ id: 'a' }, { id: 'b' }];
        SearchPageComponent.instance().onChangeSort('testSort');
        expect(SearchPageComponent.state().sortField).toBe('testSort');
        expect(utils.sortMoviesComporator).toHaveBeenCalled();
    });

    it('should set new filtered and sorted results if onChangeSort method is called', () => {
        const props = {
            movies: [{ id: 'a' }, { id: 'b' }]
        };
        const SearchPageComponentWithMovies = shallow(<SearchPage {...props} />);
        SearchPageComponentWithMovies.instance().onSearchFormSubmit('searchValue');
        expect(utils.sortMoviesComporator).toHaveBeenCalled();
        expect(utils.filterMoviesComporator).toHaveBeenCalled();
    });
});
