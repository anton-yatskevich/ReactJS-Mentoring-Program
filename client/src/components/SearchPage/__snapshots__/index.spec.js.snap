// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage component should be rendered with default props 1`] = `
<Fragment>
  <SearchPanel
    onChangeSearchField={[Function]}
    onSubmit={[Function]}
    searchField="title"
  />
  <SearchDescription
    numberOfResults={0}
    onChangeSort={[Function]}
    sortField="rating"
  />
  <ResultsList
    movies={Array []}
    onSelectMovieHandler={[Function]}
  />
</Fragment>
`;
